index.js

```diff
@@ -0,0 +1,306 @@
+
+      this.deviceInfoService = await this.server.getPrimaryService('device_information');
+      this.glucoseService = await this.server.getPrimaryService('glucose');
+      console.log('Retrieved services.');
+
+      const glucoseFeature = await this.glucoseService.getCharacteristic('glucose_feature');
+      const features = await glucoseFeature.readValue();
+      console.log('Glucose features:', features.getUint16().toString(2).padStart(8, '0'));
+
+      this.glucoseMeasurement = await this.glucoseService.getCharacteristic('glucose_measurement');
+      await this.glucoseMeasurement.startNotifications();
+      this.glucoseMeasurementContext = await this.glucoseService.getCharacteristic('glucose_measurement_context');
+      await this.glucoseMeasurementContext.startNotifications();
+      this.racp = await this.glucoseService.getCharacteristic('record_access_control_point');
+      await this.racp.startNotifications();
+      console.log('Notifications started.');
+
+      this.glucoseMeasurementContext.addEventListener('characteristicvaluechanged', bluetoothLE.handleContextNotifications);
+      this.glucoseMeasurement.addEventListener('characteristicvaluechanged', this.handleNotifications);
+      this.racp.addEventListener('characteristicvaluechanged', this.handleRACP);
+      console.log('Event listeners added.');
+    } catch (error) {
+      console.log(`Argh! ${error}`);
+      throw error;
+    }
+  }
+
+  disconnect() {
+    if (!this.device) {
+      return;
+    }
+    console.log('Disconnecting from Bluetooth Device...');
+    if (this.device.gatt.connected) {
+      this.device.gatt.disconnect();
+    } else {
+      console.log('Bluetooth Device is already disconnected');
+    }
+  }
+
+  async getDeviceInfo() {
+    console.log('Getting Device Information Characteristics...');
+    const characteristics = await this.deviceInfoService.getCharacteristics();
+    self.deviceInfo = {};
+
+    const decoder = new TextDecoder('utf-8');
+
+    /* eslint-disable no-await-in-loop, requests to devices are sequential */
+    for (let i = 0; i < characteristics.length; i += 1) {
+      switch (characteristics[i].uuid) {
+        case BluetoothUUID.getCharacteristic('manufacturer_name_string'):
+          self.deviceInfo.manufacturers = [decoder.decode(await characteristics[i].readValue())];
+          break;
+
+        case BluetoothUUID.getCharacteristic('model_number_string'):
+          self.deviceInfo.model = decoder.decode(await characteristics[i].readValue());
+          break;
+
+        default:
+          break;
+      }
+    }
+    /* eslint-enable no-await-in-loop */
+
+    return self.deviceInfo;
+  }
+
+  async sendCommand(cmd) {
+    await this.racp.writeValue(new Uint8Array(cmd));
+    console.log('Sent command.');
+  }
+
+  async getNumberOfRecords() { await this.sendCommand([0x04, 0x01]); }
+
+  async getAllRecords() {
+    self.records = [];
+    await this.sendCommand([0x01, 0x01]);
```

Comment:
Some meters I tried did not support getting the number of stored records (even though they should), so I've left this for now. We do have the capability on the backend to store the index of the last record retrieved, so we can implement this in future for meters that do support delta uploads. I've just created [UPLOAD-219] to track this.

We do not currently delete records from any meters even if supported, as some users may find this surprising/confusing.

Reply:
So how are duplicates handled?

Reply:
Deduplication is handled by the backend. As there are many devices that don't allow for only retrieving the latest records and/or deleting them, we've had deduplication built into the backend since the beginning. 

There are currently three different deduplicators on the backend, but the one that is used the most creates a hash of relevant fields in each record and uses that for deduplication.